src/Daemon/Daemon.cpp

```diff
@@ -460,6 +466,62 @@ namespace usbguard
   {
   }
 
+  pid_t Daemon::pid_fd = -1;
+
+  void Daemon::daemonize(const std::string &pid_file)
+  {
+      USBGUARD_LOG(Trace) << "Starting daemonization";
+
+      pid_t pid = 0;
+      pid_t original_pid = getpid();
+
+      sigset_t mask;
+      sigemptyset(&mask);
+      sigaddset(&mask, SIGUSR1);
+      sigprocmask(SIG_BLOCK, &mask, NULL);
+      USBGUARD_SYSCALL_THROW("Daemonize", (pid = fork()) < 0);
+      if (pid > 0) {
+        const struct timespec timeout {5,0};
+	siginfo_t info;
+	const int signum = sigtimedwait(&mask, &info, &timeout);
+	if (signum != SIGUSR1 || info.si_signo != SIGUSR1 || info.si_errno) {
```

Comment:
Here you should loop until either the SIGUSR1 is received or errno (not info.so_errno) is set to EAGAIN (meaning that the wait timed out). Other conditions, like EINTR, should be ignored.